daffodil-runtime2/src/main/scala/org/apache/daffodil/runtime2/generators/HexBinaryCodeGenerator.scala

```diff
@@ -0,0 +1,146 @@
+           | PrimType.UnsignedShort
+           | PrimType.UnsignedInt
+           | PrimType.UnsignedLong
+           | PrimType.NonNegativeInteger => "uint"
+      case p => e.SDE("Prefixed length PrimType %s is not supported in C code generator.", p.toString)
+    }
+    val intLen = e.prefixedLengthElementDecl.elementLengthInBitsEv.constValue.get
+
+    def addStatements(deref: String): Unit = {
+      val field = s"instance->$fieldName$deref"
+      val i = if (deref.length > 2) deref.substring(1, deref.length - 1) else ""
+      val lenVar = s"_l_$fieldName$i"
+      val initStatement =
+        s"""    $field.array = NULL;
+           |    $field.size = 0;
+           |    $field.dynamic = true;""".stripMargin
+      val parseStatement =
+        s"""    $intType${intLen}_t $lenVar;
+           |    parse_${conv}_$intType$intLen(&$lenVar, pstate);
+           |    if (pstate->error) return;
+           |    parse_alloc(&$field, $lenVar, pstate);
+           |    if (pstate->error) return;
+           |    parse_hexBinary(&$field, pstate);
+           |    if (pstate->error) return;""".stripMargin
+      val unparseStatement =
+        s"""    unparse_${conv}_$intType$intLen($field.size, ustate);
+           |    if (ustate->error) return;
+           |    unparse_hexBinary($field, ustate);
+           |    if (ustate->error) return;""".stripMargin
+      cgState.addSimpleTypeStatements(initStatement, parseStatement, unparseStatement)
+
+      if (e.hasFixedValue) {
+        val variable = s"${fieldName}_fixed$i"
+        val value = e.fixedValueAsString.grouped(2).mkString(", 0x")
+        val init2 = ""
+        val parse2 =
+          s"""    uint8_t $variable[] = {0x$value};
+             |    parse_validate_fixed(memcmp($field.array, $variable, sizeof($variable)) == 0, "$fieldName", pstate);
+             |    if (pstate->error) return;""".stripMargin
+        val unparse2 =
+          s"""    uint8_t $variable[] = {0x$value};
+             |    unparse_validate_fixed(memcmp($field.array, $variable, sizeof($variable)) == 0, "$fieldName", ustate);
+             |    if (ustate->error) return;""".stripMargin
+        cgState.addSimpleTypeStatements(init2, parse2, unparse2)
+      }
+    }
+    if (arraySize > 0)
+      for (i <- 0 until arraySize)
+        addStatements(s"[$i]")
+    else
+      addStatements("")
+  }
+
+  def hexBinarySpecifiedLengthGenerateCode(g: HexBinarySpecifiedLength, cgState: CodeGeneratorState): Unit = {
+    // For the time being this is a very limited back end.
+    // So there are some restrictions to enforce.
+    val e = g.e
+    e.schemaDefinitionUnless(e.bitOrder eq BitOrder.MostSignificantBitFirst, "Only dfdl:bitOrder 'mostSignificantBitFirst' is supported.")
+    e.schemaDefinitionUnless(e.elementLengthInBitsEv.isConstant, "Runtime dfdl:length expressions not supported.")
+
+    val fieldName = e.namedQName.local
+    val arraySize = if (e.occursCountKind == OccursCountKind.Fixed) e.maxOccurs else 0
```

Comment:
You might consider a different check for `arraySize`? It seems like we support either fixed length arrays or scalars. Maybe something like this instead:
```scala
val arraySize = e.occursCountKind match {
  case OccursCountKind.Fixed if e.maxOccurs > 1 => e.maxOccurs
  case OccursCountKind.Fixed if e.maxOccurs == 1 => 0
  case OccursCountKind.Implicit if e.minOccurs == 1 && e.maxOccurs == 1 => 0
  case _ => SDE("unsupported occursCountKind ...")
}
```